### ISSUE TITLE
Catching GLIBC errors in scripts

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#758]
Leading on from posts [here](http://community.mrtrix.org/t/error-response-function/393) and [here](http://community.mrtrix.org/t/error-of-running-population-template/395).

On running out of disk space (`Bus Error (core dumped)`), `libc` writes the error to the current `tty` rather than `stderr`; consequently `subprocess` can't catch it, the error disappears into the ether and people get confused.

Unfortunately the [reported solution](https://sourceware.org/ml/libc-help/2014-07/msg00008.html) doesn't work on my system: Confirmed with `VAR (std::getenv ("LIBC_FATAL_STDERR_"));` within `mrconvert` that the variable is being set, but the error still doesn't go to `stderr`. Set both at compile and run time. Baffled as to why this isn't working.

**Comment by jdtournier** [comment]
What's the exit code when this happens? MRtrix3 apps will pretty much always exit with 1 when throwing an exception - only exception to this would be if the app throws an `int` explicitly (just in case someone wanted their app to return an interpretable return code to the shell), but I don't think any of them make use of this 'feature' so far... 

So if the exit code on bus error, segfault, etc is not 0 or 1, that's a pretty clear indication of a fatal (unhandled) error. 

Other option is to [override the relevant signal handlers] (http://www.yolinux.com/TUTORIALS/C++Signals.html#DESCRIPTION) and write whatever error message we deem appropriate ourselves - and return whatever exit code we feel is appropriate so that the invoking script can also report something appropriate. Could also use this to clean up temp files (e.g. in pipes) after a ctrl-c user interrupt, segfault, broken pipe or bus error - something I've been thinking about doing for a while... Could probably set this up in `App::init()` relatively easily. Not sure how portable it would be on Windows, but I'm guessing MinGW would be fine with this. What do you reckon?

**Comment by Lestropie** [comment]
> What's the exit code when this happens?

`subprocess` is reporting a return code of `-7`. The `try{}` clause around `run()` in `command.h` isn't catching it. `7` is bus error, sounds like the negative is some kind of historical 'severity' thing.

> Other option is to override the relevant signal handlers and write whatever error message we deem appropriate ourselves - and return whatever exit code we feel is appropriate so that the invoking script can also report something appropriate.

I think this is worth pursuing: even if we catch the bus error, the current message is pretty cryptic; and if we handle such errors in the binaries, the script library may not need to be touched.

Would probably just pass whatever int code is thrown through to the process return code; means whatever is calling the binary can deal with whichever cases it wants to, and those codes are not specific to MRtrix commands.

> Could probably set this up in `App::init()`

That would make it not applicable to GUI commands; I was thinking an additional function to be called directly by `main()`.

**Comment by Lestropie** [comment]
Had a quick go at this in b331812:

* With single-threading, the intended error (as implemented in `lib/signals.h`) appears once, and the integer return code is thrown successfully (i.e. makes it to [here](https://github.com/MRtrix3/mrtrix3/blob/b331812e1beccb598779bf05a303aff417ad75ad/lib/command.h#L99). It also updates any existing ProgressBar to 100%, which may be misleading, but I'm sure that could be solved.

* With multi-threading, error appears multiple times, then get '`terminate called without an exception`' written to stderr, and '`Aborted (core dumped)`' written to tty.

* When calling an MRtrix binary within one of the Python scripts, if multi-threading is disabled, I get the appropriate error, then '`terminate called after throwing an instance of 'int'`'; with multi-threading enabled, I get multiple instances of the intended error, '`terminate called after throwing an instance of 'int'`' and '`terminate called recursively`'.

So overall, far from ideal. But maybe it's a starting point, and much of the remaining problems are due to exceptions when multi-threading (e.g. #556) more so than GLIBC. [This](http://en.cppreference.com/w/cpp/error/exception_ptr) might be a mechanism we can now embed into the multi-threading queue code in order to [propagate exceptions](http://stackoverflow.com/questions/233127/how-can-i-propagate-exceptions-between-threads) back to the `main()` thread...

**Comment by jdtournier** [comment]
Ah, no you can't throw any exceptions in these handlers, they're not regular errors, these are fatal, you have to assume the whole thing is fubar, you can't return control to the rest of the code like this (which throwing an exception will do, it'll try to unwind the stack and invoke all the destructors on its way). All you can sanely do is report the error and `exit()` immediately (or `abort()`, probably better). 

To handle the cleanup of temporary files, we'll need to maintain a list of files to be deleted, hope it's not been corrupted by whatever error occurred, and delete them one by one before aborting proper. But the main thing is to _not_ return control to the main program...

**Comment by Lestropie** [comment]
Thought that was supposed to be the case; why I was confused when the single-threaded case succeeded in making it back to be caught in `main()`. Suppose it's an undefined behavior thing.

Are there actually any temporary files other than image pipes that would need to be cleaned up?

If you think the `std::exception_ptr` is worth pursuing, can discuss in #556.

**Comment by jdtournier** [comment]
Not sure I'm totally following. The code you pushed in b331812 throws an exception of type `int`, which returns execution to the application - it doesn't matter that it's not an `Exception` as such. A signal handler should really just bail out ASAP - no throw, no return, just `abort` or `exit()`. And anything that the handler does should rely as little as possible on the rest of the address space, since when these signals are raised, execution is likely to be in a totally inconsistent state.

By the way, `App::init()` is invoked for GUI apps too ([at this point](https://github.com/MRtrix3/mrtrix3/blob/master/src/gui/gui.h#L70)). Should be perfectly safe to add in there.

One more comment: I'd avoid using a dynamic `std::map` to hold the error messages, it increases the scope for screwups in the face of likely memory corruption. Better to use a static C array, or better a switch statement in the handler itself, since then all these structures will be in the static read-only data section of the executable - no scope for corruption...

And yes, pretty sure at the moment it's only the pipes images that would need to be deleted. 

Not sure what you mean re `std::exception_ptr`, but in any case, there'll be no interaction with multi-threading if the handler exits or aborts.

**Comment by Lestropie** [comment]
> The code you pushed in b331812 ...

The commit was just plain daft. I'll have a fresh look tomorrow.

> Not sure what you mean re `std::exception_ptr`, but in any case, there'll be no interaction with multi-threading if the handler exits or aborts.

As in, for the sake of standard thrown exceptions when running multi-threaded - hence porting discussion to #556 since it's not actually relevant here.

**Comment by Lestropie** [comment]
Second attempt at this in f9d690b. Haven't been able to test because msys2 appears to not provide implementations of some of the relevant functions that it's supposed to.

Although the symbols for many signals are standardised, the actual integer values are not, and some may or may not be defined on different systems. Here I use the configure script to compile and run code that self-generates a static table of relevant terminal messages. But if it turns out that `extern const` data doesn't actually get put into the `.data` section of the executable (I couldn't find a definitive answer to this), it may be best to just use code branching instead of a lookup.

We'll also have to see if [`std::quick_exit`](http://en.cppreference.com/w/cpp/utility/program/quick_exit) does what it's supposed to: it's not actually defined on msys2 :-/

**Comment by jdtournier** [comment]
Sounds like this is a bit more complex than I'd have thought was strictly necessary...

I'd rely on pure C functionality, it's more likely to be properly supported cross-platform. So that would require only one handler function, something like this:

```C++
void signal_callback_handler (int signum)
{
  switch (signum) {
    case SIGINT:
      printf ("%s: User interrupt - exiting", MR::App::NAME.c_str());
      break;
    case SIGBUS:
      printf ("%s: Bus error: system may be running out of RAM - exiting", MR::App::NAME.c_str());
      break;
   ...
   case SIGABRT:
      printf ("%s: received abort signal - aborting", MR::App::NAME.c_str());
      abort();
  }

  // delete any files registered for deletion here

  exit (signum);
}
``` 

Note I'm trying to preserve the standard behaviour here - so SIGABRT will lead to `abort()` being called. 

Then in `App::init()`, we'd just need to register the handler for those signals we're actually interested in catching:
```C++
   signal (SIGINT, signal_callback_handler);
   signal (SIGBUS, signal_callback_handler);
   signal (SIGABRT, signal_callback_handler);
  ...
```

Then we'd need to have the facilities for marking files for deletion, which I note you've already included. I'd just have a `std::vector<std::string>` here, so we can push filenames onto it in the pipe handler (and potentially anywhere else if relevant), and just iterate through the list in the signal handler. Again, nothing fancy, the only thing I'm wondering is whether we should also have the facility to remove entries from the list, for example once the pipe handler has actually deleted its file. But if we make sure that the signal handler doesn't error out if the file it's trying to delete doesn't exist, I reckon that would probably not be necessary anyway... 

What do you reckon? That way there's no need for these massive modifications to `configure`, and we minimise any interactions with the existing code...

**Comment by Lestropie** [comment]
The main reason for doing it this way is that each of the various signal symbols is not guaranteed to be defined. Therefore you have to wrap each in its own individual `#ifdef`. If you separate out the handler setup code from the handler function code like you suggested, you end up having to maintain two long lists of `#ifdef`s; this way you only have to manage that list in one location. Not that it's a huge deal, but it's just one more way for things to go awry if you don't manage them correctly.

> `exit (signum);`

Would have thought you wouldn't want to call `exit()` since that includes cleanup. The advantage of `quick_exit()` is that you can perform limited selective cleanup, and still pass a process termination flag (unlike `abort()`). Or do you think cleaning up statics only with `exit()` should be safe?

> `signal (SIGINT, signal_callback_handler);`

The recommendation seems to be to use `sigaction()` rather than `signal()`. 

> I'd just have a `std::vector<std::string>` here

Wanted to avoid any unnecessary use of dynamic memory allocation: it's just one more de-reference that may or may not work if things have been well-and-truly corrupted. But that might lose out for the sake of being able to add any type of temporary file to this list without modification of the library code. Functionality for removal of existing entries could be done by putting it in a wrapper class and adding a mutex.

**Comment by jdtournier** [comment]
> The main reason for doing it this way is that each of the various signal symbols is not guaranteed to be defined. 

Ah. That's not cool. It does indeed make it a bit messy... Still, if it's something that can be dealt with by testing for the existence of the relevant macros, e.g.:
```C++
#ifdef SIGKILL
...
#endif
```
then I'd rather use that than add further checks to the configure script. Things could be kept relatively tidy by having a function to set up the handlers, as well as the handler itself, both defined in the same file. Obviously, if these are not macros on all systems, then there's probably no other way around.

> Would have thought you wouldn't want to call exit() since that includes cleanup.

Well, there's really not a lot at global scope in _MRtrix3_, so I don't think it makes much difference either way. But looking at the problem more closely, it seems to me that it would be simpler / cleaner to have the deletion of the temporaries done within the signal handler itself, and then call [`_Exit()`](http://en.cppreference.com/w/c/program/_Exit) (which performs _no_ cleanup whatsoever). The disadvantage being that we don't get these files deleted if the program exits due to a call to `std::exit()` or `std::quick_exit()` - but then these calls should _not_ be used in _MRtrix3_ apps, all error handling should be done via exceptions, so I don't see this being a problem. 

But this is a minor detail, happy enough with your solution, it just might be simpler if portability turns out to be a problem (`_Exit()` is C99). 

>The recommendation seems to be to use sigaction() rather than signal().

OK, happy enough with that.

> Wanted to avoid any unnecessary use of dynamic memory allocation: it's just one more de-reference that may or may not work if things have been well-and-truly corrupted. 

I get your point. But it does seem a shame to explicitly prevent this flexibility in future. In terms of these structures getting corrupted, I guess this is a very low probability event, given that they're likely to be a long way from any data actively being used at the time the signal is raised. The probability of these particular structures being affected would be small, and probably negligible when you consider how frequent fatal signals are in practice. I'd say it's probably worth having the flexibility, we don't currently do any such cleanup, and it's not been a problem so far. A failure to clean up in very rare exceptional cases is unlikely to raise any eyebrows...
 
> Functionality for removal of existing entries could be done by putting it in a wrapper class and adding a mutex.

If you want to do it right, yes. Might be enough to document that it's not thread-safe for now though - until we find some other code that might want to make use of this outside the pipe handler...

**Comment by Lestropie** [comment]
> I'd rather use that than add further checks to the configure script. 

Changed to this. It's nowhere near as elegant - in fact it looks ugly as hell in comparison - but it means e.g. addition of new signals doesn't require re-running `configure`. We could also start handling different signals in different ways: e.g. looks like `SIGBUS` occurs when the memory-mapped region gets truncated to fit into the available disk space; so technically we could probably do a full cleanup in that case.

I also switched the functionality toggle to `#ifdef MRTRIX_SIGNAL_HANDLING` rather than `#ifndef MRTRIX_NO_SIGNAL_HANDLING`. That way if a system doesn't support this functionality (e.g. my laptop), and fails to re-run `configure`, MRtrix should still compile. But it means most users will need to re-run `configure` before the functionality kicks in.

> But looking at the problem more closely, it seems to me that it would be simpler / cleaner to have the deletion of the temporaries done within the signal handler itself, and then call _Exit() (which performs no cleanup whatsoever). The disadvantage being that we don't get these files deleted if the program exits due to a call to std::exit() or std::quick_exit() - but then these calls should not be used in MRtrix3 apps, all error handling should be done via exceptions, so I don't see this being a problem.

All `std::quick_exit()` does is call the cleanup functions registered through `std::at_quick_exit()`, then call `_Exit()`; so it's basically the same as what you're suggesting. But I've changed to an explicit `std::_Exit()` call on the off chance that a dev might see a `std::on_quick_exit()` and use it where an exception should be instead thrown. If more sophisticated cleanup on system signal is one day required (i.e. over and above a list of filename strings, thereby requiring callback functions), we can redesign.

> I'd just have a std::vector<std::string> here, so we can push filenames onto it in the pipe handler (and potentially anywhere else if relevant), and just iterate through the list in the signal handler.

Done.
